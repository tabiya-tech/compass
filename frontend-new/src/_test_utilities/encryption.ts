import { webcrypto as crypto } from "crypto";
import { EncryptionConfig } from "src/sensitiveData/config/encryptionConfig";

// Constraining the key size to 2048 or 4096 bits for RSA as the backend expect the 4096 in calculating the maximum size of the encrypted key.
export async function generateRSACryptoPairKey(keySize: 2048 | 4096) {
  return await crypto.subtle.generateKey(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
      modulusLength: keySize,
      // 3 is a common exponent in practice
      publicExponent: new Uint8Array([1, 0, 1]),
      hash: EncryptionConfig.RSA.KEY_HASH_FN,
    },
    true, // Whether the key is extractable (i.e. can be used in exportKey)
    ["encrypt", "decrypt"] // the key can be used for both encryption and decryption
  );
}

function arrayBufferToBase64(buffer: ArrayBuffer): string {
  // @ts-ignore
  const binary = String.fromCharCode(...new Uint8Array(buffer));

  return Buffer.from(binary, "binary").toString("base64");
}

/**
 * Export a CryptoKey to a PEM formatted string (public key)
 *
 * @param key The CryptoKey to export as a PEM formatted string
 * @returns The PEM formatted string.
 */
export async function exportCryptoPublicKey(key: CryptoKey) {
  // Export the key to ArrayBuffer
  const exported = await crypto.subtle.exportKey(EncryptionConfig.RSA.FORMAT, key);

  const base64String = arrayBufferToBase64(exported);

  // add the PEM header and footer to make it a PEM formatted string.
  return `-----BEGIN PUBLIC KEY-----\n${base64String}\n-----END PUBLIC KEY-----`;
}

/**
 * Decrypt the encrypted data with the private key
 * using RSA Algorithm.
 *
 * @param privateKey
 * @param encryptedData
 */
export async function decryptWithRSA(privateKey: CryptoKey, encryptedData: ArrayBuffer) {
  return await crypto.subtle.decrypt(
    {
      name: EncryptionConfig.RSA.ALGORITHM,
    },
    privateKey,
    encryptedData
  );
}

export async function decryptWithAES(encryptedData: ArrayBuffer, iv: ArrayBuffer, key: ArrayBuffer) {
  try {
    const importedKey = await crypto.subtle.importKey(
      EncryptionConfig.AES.FORMAT,
      key,
      EncryptionConfig.AES.ALGORITHM,
      true,
      ["decrypt"]
    );

    return await crypto.subtle.decrypt(
      {
        name: EncryptionConfig.AES.ALGORITHM,
        iv,
        length: EncryptionConfig.AES.KEY_LEN,
        tagLength: EncryptionConfig.AES.TAG_LEN,
      },
      importedKey,
      encryptedData
    );
  } catch (e) {
    console.error(e);
    throw e;
  }
}
