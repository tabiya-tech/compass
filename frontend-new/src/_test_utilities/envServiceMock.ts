jest.mock("src/envService", () => ({
  getFirebaseAPIKey: jest.fn(() => "mock-api-key"),
  getFirebaseDomain: jest.fn(() => "mock-auth-domain"),
  getBackendUrl: jest.fn(() => "mock-backend-url"),
  getApplicationLoginCode: jest.fn(() => ""),
  getApplicationRegistrationCode: jest.fn(() => ""),
  getLoginCodeDisabled: jest.fn(() => "false"),
  getRegistrationDisabled: jest.fn(() => "false"),
  getMetricsEnabled: jest.fn(() => "true"),
  getMetricsConfig: jest.fn(() => ""),
  getCvUploadEnabled: jest.fn(() => "true"),
  getSocialAuthDisabled: jest.fn(() => "false"),
}));
