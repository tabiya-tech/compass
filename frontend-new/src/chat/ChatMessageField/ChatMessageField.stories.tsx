import React, { useEffect, useState } from "react";
import type { Meta, StoryObj } from "@storybook/react";
import ChatMessageField, { ChatMessageFieldProps } from "./ChatMessageField";
import { action } from "@storybook/addon-actions";

const meta: Meta<typeof ChatMessageField> = {
  title: "Chat/ChatMessageField",
  component: ChatMessageField,
  tags: ["autodocs"],
  argTypes: {
    handleSend: { action: "message sent" },
    notifyChange: { action: "change happened" },
  },
};

export default meta;

type Story = StoryObj<typeof ChatMessageField>;

const ChatMessageFieldWrapper: React.FC<ChatMessageFieldProps> = (props) => {
  const [message, setMessage] = useState<string>(props.message);

  const handleSend = () => {
    props.handleSend();
    setMessage("");
  };

  useEffect(() => {
    if (message.length > 0) props.notifyChange(message);
  }, [message, props]);

  return <ChatMessageField {...props} message={message} notifyChange={setMessage} handleSend={handleSend} />;
};

export const Shown: Story = {
  render: (args) => <ChatMessageFieldWrapper {...args} />,
  args: {
    message: "",
    handleSend: action("Message sent"),
  },
};

export const AIIsTyping: Story = {
  render: (args) => <ChatMessageFieldWrapper {...args} />,
  args: {
    message: "",
    handleSend: action("Message sent"),
    aiIsTyping: true,
  },
};

export const ChatIsClosed: Story = {
  render: (args) => <ChatMessageFieldWrapper {...args} />,
  args: {
    message: "",
    handleSend: action("Message sent"),
    isChatFinished: true,
  },
};
