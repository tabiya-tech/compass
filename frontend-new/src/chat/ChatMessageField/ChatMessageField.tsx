import React, { useEffect, useContext, useMemo, useState, MouseEvent } from "react";
import { IconButton, InputAdornment, TextField, styled, useTheme, Typography, Box } from "@mui/material";
import SendIcon from "@mui/icons-material/Send";
import { IsOnlineContext } from "src/app/isOnlineProvider/IsOnlineProvider";

export interface ChatMessageFieldProps {
  message: string;
  notifyChange: React.Dispatch<React.SetStateAction<string>>;
  handleSend: () => void;
  aiIsTyping: boolean;
  isChatFinished: boolean;
}

const uniqueId = "2a76494f-351d-409d-ba58-e1b2cfaf2a53";
export const DATA_TEST_ID = {
  CHAT_MESSAGE_FIELD_CONTAINER: `chat-message-field-container-${uniqueId}`,
  CHAT_MESSAGE_FIELD: `chat-message-field-${uniqueId}`,
  CHAT_MESSAGE_FIELD_BUTTON: `chat-message-field-button-${uniqueId}`,
  CHAT_MESSAGE_FIELD_ICON: `chat-message-field-icon-${uniqueId}`,
  CHAT_MESSAGE_CHAR_COUNTER: `chat-message-char-counter-${uniqueId}`,
};

export const CHAT_MESSAGE_MAX_LENGTH = 1000;
export const DISALLOWED_CHARACTERS = /["\\{}[\]*_#`<>~|]/g;

const StyledTextField = styled(TextField)(({ theme, disabled }) => ({
  "& .MuiOutlinedInput-root": {
    "& fieldset": {
      borderColor: theme.palette.primary.main,
      borderWidth: theme.fixedSpacing(theme.tabiyaSpacing.xxs),
    },
    "&:hover fieldset": {
      borderColor: !disabled ? theme.palette.primary.dark : theme.palette.grey[400],
    },
    "&.Mui-focused fieldset": {
      borderColor: theme.palette.primary.dark,
    },
    "&.Mui-error fieldset": {
      borderColor: theme.palette.error.main,
    },
  },

  "& .css-111mkcg-MuiInputBase-input-MuiOutlinedInput-input": {
    paddingRight: theme.spacing(4),

    "&.Mui-disabled": {
      WebkitTextFillColor: theme.palette.text.textBlack,
    },
  },
}));

const ChatMessageField: React.FC<ChatMessageFieldProps> = (props) => {
  const inputRef = React.useRef<HTMLInputElement>(null);

  const theme = useTheme();
  const [errorMessage, setErrorMessage] = useState<string>("");
  const [charCount, setCharCount] = useState<number>(0);

  const isOnline = useContext(IsOnlineContext);

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const inputValue = e.target.value;
    let errorMessage = "";

    // Filter out disallowed characters
    let filteredValue = inputValue.replace(DISALLOWED_CHARACTERS, "");
    const invalidChar = inputValue.split("").filter((char) => DISALLOWED_CHARACTERS.test(char));

    // Update character count
    if (filteredValue) {
      setCharCount(filteredValue.length);
    }

    // Check for character limit
    if (filteredValue.length >= CHAT_MESSAGE_MAX_LENGTH) {
      errorMessage = "Message limit is 1000 characters.";
    }

    // Check for special characters in original input
    if (inputValue !== filteredValue) {
      errorMessage = `Invalid special characters: ${invalidChar}`;
    }

    setErrorMessage(errorMessage);

    if (props.notifyChange) {
      props.notifyChange(filteredValue);
    }
  };

  const handleButtonClick = (e: MouseEvent) => {
    e.stopPropagation();

    // Prevent sending a message when the AI is typing, or a message is still being sent.
    if (props.aiIsTyping) return;

    // Prevent sending a message when the chat is finished.
    if (props?.isChatFinished) return;

    // Prevent sending an empty message
    if (!props.message.length) return;

    props.handleSend();
    setErrorMessage("");
    setCharCount(0);
  };

  // The character counter is only visible when the user types more than 75% of the character limit
  const showCharCounter = charCount > CHAT_MESSAGE_MAX_LENGTH * 0.75;
  // The character counter turns red when only 1% is left
  const counterColor =
    charCount >= CHAT_MESSAGE_MAX_LENGTH * 0.95 ? theme.palette.error.main : theme.palette.text.secondary;

  const placeHolder = useMemo(() => {
    if (props.isChatFinished) {
      return "Conversation has been completed. You can't send any more messages.";
    }
    if (props.aiIsTyping) {
      return "AI is typing..., wait for it to finish.";
    }
    if (!isOnline) {
      return "You are offline. Please connect to the internet to send a message.";
    }
    return "Type your message...";
  }, [props.aiIsTyping, props.isChatFinished, isOnline]);

  const isDisabled = props.isChatFinished || props.aiIsTyping;

  useEffect(() => {
    if (inputRef.current && !props.aiIsTyping) {
      inputRef.current.focus();
    }
  }, [props.aiIsTyping]);

  // if the input is focused and the user scrolls, the input should be blurred
  useEffect(() => {
    const handleTouchEnd = () => {
      // if the input is focused, blur it
      if (document.activeElement === inputRef.current) {
        inputRef.current?.blur();
      }
    };
    window.addEventListener("touchend", handleTouchEnd);
    return () => {
      window.removeEventListener("touchend", handleTouchEnd);
    };
  }, []);

  return (
    <Box
      position="relative"
      data-testid={DATA_TEST_ID.CHAT_MESSAGE_FIELD_CONTAINER}
      sx={{
        width: "100%",
        alignItems: "center",
        display: "flex",
        flexDirection: "column",
      }}
    >
      <Box
        sx={{
          width: {
            xs: "100%",
            md: "60%",
          },
          position: "relative",
        }}
      >
        <StyledTextField
          placeholder={placeHolder}
          variant="outlined"
          fullWidth
          multiline
          maxRows={4}
          value={props.message}
          disabled={isDisabled}
          onChange={handleChange}
          inputRef={inputRef}
          error={!!errorMessage}
          helperText={errorMessage}
          InputProps={{
            endAdornment: (
              <InputAdornment position="end">
                <IconButton
                  data-testid={DATA_TEST_ID.CHAT_MESSAGE_FIELD_BUTTON}
                  onClick={handleButtonClick}
                  disabled={isDisabled || !isOnline}
                  color={"error"}
                  title="send message"
                >
                  <SendIcon
                    data-testid={DATA_TEST_ID.CHAT_MESSAGE_FIELD_ICON}
                    sx={{
                      color: isDisabled || !isOnline ? theme.palette.grey[400] : theme.palette.primary.dark,
                    }}
                  />
                </IconButton>
              </InputAdornment>
            ),
            inputProps: {
              maxLength: 1000,
              "data-testid": DATA_TEST_ID.CHAT_MESSAGE_FIELD,
            },
          }}
        />
        {showCharCounter && (
          <Typography
            data-testid={DATA_TEST_ID.CHAT_MESSAGE_CHAR_COUNTER}
            variant="caption"
            color={counterColor}
            sx={{ position: "absolute", bottom: 0, right: 10 }}
          >
            {charCount}/{CHAT_MESSAGE_MAX_LENGTH}
          </Typography>
        )}
      </Box>
    </Box>
  );
};

export default ChatMessageField;
