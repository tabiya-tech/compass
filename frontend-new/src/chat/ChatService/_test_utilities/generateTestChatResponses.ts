import { ConversationMessage, ConversationMessageSender } from "src/chat/ChatService/ChatService.types";

export const generateTestChatResponses = (): ConversationMessage[] => {
  return [
    {
      message: "Hello! How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "I can help you with that. What is your name?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
  ];
};

export const generateTestHistory = (): ConversationMessage[] => {
  return [
    {
      message: "Hello! How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "I can help you with that. What is your name?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "My name is John.",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.USER,
    },
    {
      message: "Nice to meet you, John. How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
  ];
};
