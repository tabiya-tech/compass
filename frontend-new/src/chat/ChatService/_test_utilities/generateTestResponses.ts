import { LastMessage, RootObject } from "src/chat/ChatService/ChatService.types";

export const generateRootObjectResponse = (): RootObject => ({
  last: {
    message_for_user: "Hello",
    finished: false,
    agent_type: "WelcomeAgent",
    reasoning: "Initial greeting",
    agent_response_time_in_sec: 1.0,
    llm_stats: [
      {
        error: "",
        prompt_token_count: 50,
        response_token_count: 10,
        response_time_in_sec: 1.0,
      },
    ],
  },
  conversation_context: {
    all_history: {
      turns: [
        {
          index: 1,
          input: {
            message: "Hello",
          },
          output: {
            message_for_user: "Hi there!",
            finished: false,
            agent_type: "WelcomeAgent",
            reasoning: "Greeting response",
            agent_response_time_in_sec: 0.5,
            llm_stats: [
              {
                error: "",
                prompt_token_count: 30,
                response_token_count: 10,
                response_time_in_sec: 0.5,
              },
            ],
          },
        },
      ],
    },
    history: {
      turns: [],
    },
    summary: "",
  },
});

export const generateLastMessageResponse = (): LastMessage => ({
  message_for_user: "Hi there!",
  finished: false,
  agent_type: "WelcomeAgent",
  reasoning: "Greeting response",
  agent_response_time_in_sec: 1.0,
  llm_stats: [
    {
      error: "",
      prompt_token_count: 30,
      response_token_count: 10,
      response_time_in_sec: 0.5,
    },
  ],
});
