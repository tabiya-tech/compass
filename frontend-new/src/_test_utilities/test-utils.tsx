// Based on https://testing-library.com/docs/react-testing-library/setup/
import React, { ReactElement, ReactNode } from "react";
import { render, renderHook, RenderHookOptions, RenderOptions } from "@testing-library/react";
import { ThemeProvider } from "@mui/material";
import applicationTheme, { ThemeMode } from "src/theme/applicationTheme/applicationTheme";

// Import the Firebase mock utilities
import "src/_test_utilities/firebaseMock";
import SnackbarProvider from "src/theme/SnackbarProvider/SnackbarProvider";
import { IsOnlineProvider } from "src/app/isOnlineProvider/IsOnlineProvider";
jest.mock("firebase/compat/app", () => {
  return {
    initializeApp: jest.fn(),
    auth: jest.fn().mockReturnValue({
      signOut: jest.fn(),
      currentUser: {
        getIdToken: jest.fn(),
        getTokenResult: jest.fn(),
      },
      onAuthStateChanged: jest.fn().mockReturnValue(jest.fn()),
      signInWithPopup: jest.fn(),
      GoogleAuthProvider: { PROVIDER_ID: "google.com" },
    }),
  };
});

jest.mock("firebaseui", () => {
  return {
    auth: {
      AuthUI: {
        start: jest.fn(),
        getInstance: jest.fn().mockReturnValue({
          start: jest.fn(),
          reset: jest.fn(),
        }),
        reset: jest.fn(),
      },
    },
  };
});

jest.mock("@sentry/react", () => ({
  withSentry: (Component: any) => Component,
  getFeedback: jest.fn(),
  withProfiler: (Component: any) => Component,
  wrapCreateBrowserRouter: (Component: any) => Component,
}));

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <IsOnlineProvider>
      <ThemeProvider theme={applicationTheme(ThemeMode.LIGHT)}>
        <SnackbarProvider>{children}</SnackbarProvider>
      </ThemeProvider>
    </IsOnlineProvider>
  );
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, { wrapper: AllTheProviders, ...options });

function customRenderHook<T>(
  hook: () => T,
  options?: Omit<RenderHookOptions<T>, "wrapper"> & { wrapper?: React.ComponentType }
) {
  const CombinedWrapper = ({ children }: { children: ReactNode }) => {
    const Wrapper = options?.wrapper ?? React.Fragment;
    return (
      <AllTheProviders>
        <Wrapper>{children}</Wrapper>
      </AllTheProviders>
    );
  };

  return renderHook(hook, { wrapper: CombinedWrapper, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
export { customRenderHook as renderHook };
