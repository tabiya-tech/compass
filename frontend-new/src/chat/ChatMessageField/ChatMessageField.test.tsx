// mute the console
import "src/_test_utilities/consoleMock";

import ChatMessageField, { DATA_TEST_ID, CHAT_MESSAGE_MAX_LENGTH, DISALLOWED_CHARACTERS } from "./ChatMessageField";
import { render, screen, fireEvent } from "src/_test_utilities/test-utils";
import { mockBrowserIsOnLine } from "src/_test_utilities/mockBrowserIsOnline";

jest.mock("src/auth/services/socialAuth/SocialAuth.service", () => {
  return {
    __esModule: true,
    default: jest.fn().mockImplementation(() => {
      return {
        logout: jest.fn(),
      };
    }),
  };
});

describe("ChatMessageField", () => {
  test("should render ChatMessageField correctly", () => {
    // WHEN ChatMessageField is rendered
    render(
      <ChatMessageField
        aiIsTyping={false}
        isChatFinished={false}
        handleSend={jest.fn()}
        message="foo"
        notifyChange={jest.fn()}
      />
    );

    //THEN expect no errors or warnings has occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();
    // AND the ChatMessageField container to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD_CONTAINER)).toBeInTheDocument();
    // AND the ChatMessageField input to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD)).toBeInTheDocument();
    // AND the ChatMessageField button to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD_BUTTON)).toBeInTheDocument();
    // AND the ChatMessageField icon to be in the document
    expect(screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD_ICON)).toBeInTheDocument();
    // AND to match the snapshot
    expect(screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD_CONTAINER)).toMatchSnapshot();
  });

  test("should render ChatMessageField correctly with error message when message is too long", () => {
    // GIVEN a long message
    const message = "a".repeat(CHAT_MESSAGE_MAX_LENGTH + 2);
    // AND handleChange function
    const handleChange = jest.fn();

    // WHEN ChatMessageField is rendered
    render(
      <ChatMessageField
        aiIsTyping={false}
        isChatFinished={false}
        handleSend={jest.fn()}
        message=""
        notifyChange={handleChange}
      />
    );
    // AND the input is changed
    const ChatMessageFieldInput = screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD);
    fireEvent.change(ChatMessageFieldInput, { target: { value: message } });

    // THEN expect notifyChange to be called
    expect(handleChange).toHaveBeenCalled();
    // AND the error message to be in the document
    expect(screen.getByText("Message limit is 1000 characters.")).toBeInTheDocument();
  });

  test("should render ChatMessageField correctly with error message when message contains invalid characters", () => {
    // WHEN a message that contains invalid characters
    const invalidMessage = "foobar{}&*";
    const invalidChar = invalidMessage.split("").filter((char) => DISALLOWED_CHARACTERS.test(char));
    // AND handleChange function
    const handleChange = jest.fn();

    // WHEN ChatMessageField is rendered
    render(
      <ChatMessageField
        aiIsTyping={false}
        isChatFinished={false}
        handleSend={jest.fn()}
        message=""
        notifyChange={handleChange}
      />
    );
    // AND the input is changed
    const ChatMessageFieldInput = screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD);
    fireEvent.change(ChatMessageFieldInput, { target: { value: invalidMessage } });

    // THEN expect notifyChange to be called
    expect(handleChange).toHaveBeenCalled();
    // AND the error message to be in the document
    expect(screen.getByText(`Invalid special characters: ${invalidChar}`)).toBeInTheDocument();
  });

  test("should render ChatMessageField correctly with no error message when message is valid", () => {
    // GIVEN a valid message
    const validMessage = "foo bar";
    // AND handleChange function
    const handleChange = jest.fn();

    // WHEN ChatMessageField is rendered
    render(
      <ChatMessageField
        aiIsTyping={false}
        isChatFinished={false}
        handleSend={jest.fn()}
        message=""
        notifyChange={handleChange}
      />
    );
    // AND the input is changed
    const ChatMessageFieldInput = screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD);
    fireEvent.change(ChatMessageFieldInput, { target: { value: validMessage } });

    // THEN expect notifyChange to be called
    expect(handleChange).toHaveBeenCalled();
    // AND no error message to be in the document
    expect(screen.queryByText("Message limit is 1000 characters.")).toBeNull();
    expect(screen.queryByText("Invalid special characters.")).toBeNull();
  });

  describe("character counter", () => {
    test("should render Character counter when the message is longer than 75% of the max allowed length", () => {
      // GIVEN a long message (76%)
      const message = "a".repeat(CHAT_MESSAGE_MAX_LENGTH * 0.76);

      // AND handleChange function
      const handleChange = jest.fn();

      // WHEN ChatMessageField is rendered
      render(
        <ChatMessageField
          aiIsTyping={false}
          isChatFinished={false}
          handleSend={jest.fn()}
          message=""
          notifyChange={handleChange}
        />
      );

      // AND the input is changed
      const ChatMessageFieldInput = screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD);

      fireEvent.change(ChatMessageFieldInput, { target: { value: message } });

      // THEN expect the character counter to be in the document
      expect(screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_CHAR_COUNTER)).toBeInTheDocument();
      // AND the character counter to be correct
      expect(screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_CHAR_COUNTER)).toHaveTextContent(
        `${message.length}/${CHAT_MESSAGE_MAX_LENGTH}`
      );
    });

    test("should not render the character counter when the message is shorter than or equal to 75% of the max allowed length", () => {
      // GIVEN a short message (75%)
      const message = "a".repeat(CHAT_MESSAGE_MAX_LENGTH * 0.75);

      // AND handleChange function
      const handleChange = jest.fn();

      // WHEN ChatMessageField is rendered
      render(
        <ChatMessageField
          aiIsTyping={false}
          isChatFinished={false}
          handleSend={jest.fn()}
          message=""
          notifyChange={handleChange}
        />
      );

      // AND the input is changed
      const ChatMessageFieldInput = screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD);

      fireEvent.change(ChatMessageFieldInput, { target: { value: message } });

      // THEN expect the character counter not to be in the document
      expect(screen.queryByTestId(DATA_TEST_ID.CHAT_MESSAGE_CHAR_COUNTER)).not.toBeInTheDocument();
    });
  });

  describe("ChatMessageField action tests", () => {
    test("should call handleSend when button is clicked", () => {
      // GIVEN handleSend function
      const handleSend = jest.fn();

      // WHEN ChatMessageField is rendered
      render(
        <ChatMessageField
          aiIsTyping={false}
          isChatFinished={false}
          handleSend={handleSend}
          message="foo"
          notifyChange={jest.fn()}
        />
      );
      // AND the button is clicked
      const ChatMessageFieldButton = screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD_BUTTON);
      fireEvent.change(screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD), { target: { value: "foo" } });
      fireEvent.click(ChatMessageFieldButton);

      // THEN expect handleSend to be called
      expect(handleSend).toHaveBeenCalled();
    });

    test("should call notifyChange when input is changed", () => {
      // GIVEN notifyChange function
      const notifyChange = jest.fn();

      // WHEN ChatMessageField is rendered
      render(
        <ChatMessageField
          aiIsTyping={false}
          isChatFinished={false}
          handleSend={jest.fn()}
          message="foo"
          notifyChange={notifyChange}
        />
      );
      // AND the input is changed
      const ChatMessageFieldInput = screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD);
      fireEvent.change(ChatMessageFieldInput, { target: { value: "bar" } });

      // THEN expect notifyChange to be called
      expect(notifyChange).toHaveBeenCalled();
    });

    test("should disable sending a message when the message is empty", () => {
      // GIVEN handleSend function
      const handleSend = jest.fn();
      // WHEN ChatMessageField is rendered
      render(
        <ChatMessageField
          aiIsTyping={false}
          isChatFinished={false}
          handleSend={handleSend}
          message=""
          notifyChange={jest.fn()}
        />
      );
      // AND the button is clicked
      const ChatMessageFieldButton = screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD_BUTTON);
      fireEvent.click(ChatMessageFieldButton);

      // THEN expect handleSend not to be called
      expect(handleSend).not.toHaveBeenCalled();
    });

    it("should disable send button when AI is typing", () => {
      // GIVEN handleSend function
      const handleSend = jest.fn();
      // WHEN ChatMessageField is rendered
      render(
        <ChatMessageField
          aiIsTyping={true}
          isChatFinished={false}
          handleSend={handleSend}
          message="foo"
          notifyChange={jest.fn()}
        />
      );
      // AND the button is clicked
      const ChatMessageFieldButton = screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD_BUTTON);

      // THEN expect ChatMessageFieldButton to be disabled
      expect(ChatMessageFieldButton).toBeDisabled();
    });

    it("should not call handleSend when AI is typing", () => {
      // GIVEN handleSend function
      const handleSend = jest.fn();
      // WHEN ChatMessageField is rendered with some message
      render(
        <ChatMessageField
          aiIsTyping={true}
          isChatFinished={false}
          handleSend={handleSend}
          message="foo"
          notifyChange={jest.fn()}
        />
      );

      // AND the button is clicked
      const ChatMessageFieldButton = screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD_BUTTON);
      fireEvent.click(ChatMessageFieldButton);

      // THEN expect handleSend not to be called
      expect(handleSend).not.toHaveBeenCalled();
    });

    it("should be disabled when chat is finished", () => {
      // GIVEN handleSend function
      const handleSend = jest.fn();
      // WHEN ChatMessageField is rendered
      render(
        <ChatMessageField
          aiIsTyping={false}
          isChatFinished={true}
          handleSend={handleSend}
          message="foo"
          notifyChange={jest.fn()}
        />
      );
      // AND the button is clicked
      const ChatMessageFieldButton = screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD_BUTTON);

      // THEN expect ChatMessageFieldButton to be disabled
      expect(ChatMessageFieldButton).toBeDisabled();
    });

    test("should be disabled when browser is offline", () => {
      // GIVEN the browser is offline
      mockBrowserIsOnLine(false);
      // AND handleSend function
      const handleSend = jest.fn();
      // WHEN ChatMessageField is rendered
      render(
        <ChatMessageField
          aiIsTyping={false}
          isChatFinished={false}
          handleSend={handleSend}
          message="foo"
          notifyChange={jest.fn()}
        />
      );
      // AND the button is clicked
      const ChatMessageFieldButton = screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD_BUTTON);

      // THEN expect ChatMessageFieldButton to be disabled
      expect(ChatMessageFieldButton).toBeDisabled();

      // AND the button is clicked
      fireEvent.click(ChatMessageFieldButton);

      // THEN expect handleSend not to be called
      expect(handleSend).not.toHaveBeenCalled();
    });

    it("should focus on the input once ai is done typing", async () => {
      // GIVEN handleSend function
      const handleSend = jest.fn();

      // WHEN ChatMessageField is rendered
      render(
        <ChatMessageField
          aiIsTyping={false}
          isChatFinished={false}
          handleSend={handleSend}
          message="foo"
          notifyChange={jest.fn()}
        />
      );

      const messageField = screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_FIELD);

      // THEN expect input should be focused
      expect(messageField).toHaveFocus();
    });
  });
});
