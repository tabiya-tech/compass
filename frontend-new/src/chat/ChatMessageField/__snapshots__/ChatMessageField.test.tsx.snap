// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatMessageField should render ChatMessageField correctly 1`] = `
<div
  class="MuiBox-root css-acf3dn"
  data-testid="chat-message-field-container-2a76494f-351d-409d-ba58-e1b2cfaf2a53"
>
  <div
    class="MuiBox-root css-zeiu9o"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-r75dsv-MuiFormControl-root-MuiTextField-root"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl MuiInputBase-multiline MuiInputBase-adornedEnd css-8z051r-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <textarea
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputAdornedEnd css-f55ppk-MuiInputBase-input-MuiOutlinedInput-input"
          data-testid="chat-message-field-2a76494f-351d-409d-ba58-e1b2cfaf2a53"
          id=":r0:"
          maxlength="1000"
          placeholder="Type your message..."
          style="height: 0px; overflow: hidden;"
        >
          foo
        </textarea>
        <textarea
          aria-hidden="true"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiInputBase-inputAdornedEnd css-f55ppk-MuiInputBase-input-MuiOutlinedInput-input"
          readonly=""
          style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
          tabindex="-1"
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorError MuiIconButton-sizeMedium css-1fhje8j-MuiButtonBase-root-MuiIconButton-root"
            data-testid="chat-message-field-button-2a76494f-351d-409d-ba58-e1b2cfaf2a53"
            tabindex="0"
            title="send message"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-ldtitb-MuiSvgIcon-root"
              data-testid="chat-message-field-icon-2a76494f-351d-409d-ba58-e1b2cfaf2a53"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M2.01 21 23 12 2.01 3 2 10l15 2-15 2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-ihdtdm"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
